src/move/research/uniswap-v3-book-milestone-1/sources/math.move

```diff
@@ -0,0 +1,70 @@
+// cspell:words stdlib
+module math::math {
+
+    const SHIFT_Q64: u8 = 64;
+
+    inline fun sqrt(i: u256): u256 {
+        if (i <= 1) {
+            i
+        } else {
+            let estimate = 1 << ((log2_unchecked(i) / 2) + 1);
+            let update;
+            loop {
+                update = (estimate + i / estimate) / 2;
+                if (update >= estimate) break;
+                estimate = update;
+            };
+            estimate
+        }
+    }
+
+    /// Does not check for case of `i` = 0, which is undefined.
+    inline fun log2_unchecked(i: u256): u8 {
+        let result = 0;
+        let check_bit = 128;
+        loop {
+            if (i >= (1 << check_bit)) {
+                i = i >> check_bit;
+                result = result + check_bit;
+            };
+            check_bit = check_bit / 2;
+            if (check_bit == 0) break
+        };
+        result
+    }
+
+    inline fun u64_to_q64(x: u64): u128 { (x as u128) << SHIFT_Q64 }
+
+    inline fun q64_to_u64(x: u128): u64 { ((x >> SHIFT_Q64) as u64) }
+
+    inline fun sqrt_q64(x: u128): u128 { (sqrt((x as u256) << SHIFT_Q64) as u128) }
+
+    #[test]
+    /// Adapted from `aptos_stdlib::math128`.
+    fun test_log2_unchecked() {
+        let bit = 255;
+        while (bit > 0) {
+            let operand = (1 << bit) + 1;
```

Comment:
To be clear, this is testing 2^255 + 1 down to 2^1 + 1 to verify the flooring and log calculation, yea?